File: src/main/java/jp/co/cybird/barcodekanojoForGAM/adapter/KanojoInfoImgAdapter.java
package jp.co.cybird.barcodekanojoForGAM.adapter;

import android.content.Context;
import android.os.Handler;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.Gallery;
import android.widget.ImageView;
import java.util.ArrayList;
import java.util.List;
import java.util.Observable;
import java.util.Observer;
import jp.co.cybird.barcodekanojoForGAM.R;
import jp.co.cybird.barcodekanojoForGAM.adapter.base.ObservableAdapter;
import jp.co.cybird.barcodekanojoForGAM.core.util.ImageCache;
import jp.co.cybird.barcodekanojoForGAM.core.util.RemoteResourceManager;

public class KanojoInfoImgAdapter extends BaseAdapter implements ObservableAdapter {

    private static final String TAG = "KanojoInfoImgAdapter";
    private Context context;
    private int galleryIconSize;
    private List<String> imgUrls;
    private Handler mHandler;
    private final Runnable mNotifyThread = new Runnable() {
        public void run() {
            KanojoInfoImgAdapter.this.notifyDataSetChanged();
        }
    };
    private RemoteResourceManagerObserver mResourcesObserver;
    private RemoteResourceManager mRrm;

    public KanojoInfoImgAdapter(Context context2, RemoteResourceManager rrm) {
        this.context = context2;
        context2.obtainStyledAttributes(R.styleable.KanojoInfoGallery).recycle();
        this.imgUrls = new ArrayList();
        this.galleryIconSize = (int) context2.getResources().getDimension(R.dimen.kanojo_info_gallery_height);
        this.mHandler = new Handler();
        this.mRrm = rrm;
        this.mResourcesObserver = new RemoteResourceManagerObserver(this, (RemoteResourceManagerObserver) null);
        this.mRrm.addObserver(this.mResourcesObserver);
    }

    public void setImgUrls(List<String> l) {
        this.imgUrls = l;
        for (String url : l) {
            ImageCache.requestImage(url, this.mRrm);
        }
        notifyDataSetChanged();
        this.mHandler.removeCallbacks(this.mNotifyThread);
        this.mHandler.postDelayed(this.mNotifyThread, 200);
    }

    public void addImgUrl(String l) {
        if (this.imgUrls == null) {
            this.imgUrls = new ArrayList();
        }
        this.imgUrls.add(l);
        notifyDataSetChanged();
        ImageCache.requestImage(l, this.mRrm);
        this.mHandler.removeCallbacks(this.mNotifyThread);
        this.mHandler.postDelayed(this.mNotifyThread, 200);
    }

    public void clear() {
        this.imgUrls.clear();
    }

    public int getCount() {
        if (this.imgUrls == null) {
            return 0;
        }
        return this.imgUrls.size();
    }

    public Object getItem(int position) {
        return this.imgUrls.get(position);
    }

    public boolean hasStableIds() {
        return true;
    }

    public String getItemUrl(int position) {
        return this.imgUrls.get(position);
    }

    public long getItemId(int position) {
        return (long) position;
    }

    public View getView(int position, View view, ViewGroup parent) {
        ImageView imageView;
        if (view == null) {
            imageView = new ImageView(this.context);
            imageView.setScaleType(ImageView.ScaleType.FIT_CENTER);
            imageView.setLayoutParams(new Gallery.LayoutParams(this.galleryIconSize, this.galleryIconSize));
        } else {
            imageView = (ImageView) view;
        }
        ImageCache.setImage(imageView, this.imgUrls.get(position), this.mRrm, R.drawable.common_noimage_product);
        return imageView;
    }

    public void removeObserver() {
        this.mHandler.removeCallbacks(this.mNotifyThread);
        this.mRrm.deleteObserver(this.mResourcesObserver);
    }

    private class RemoteResourceManagerObserver implements Observer {
        private RemoteResourceManagerObserver() {
        }

        /* synthetic */ RemoteResourceManagerObserver(KanojoInfoImgAdapter kanojoInfoImgAdapter, RemoteResourceManagerObserver remoteResourceManagerObserver) {
            this();
        }

        public void update(Observable observable, Object data) {
            if (data != null) {
                KanojoInfoImgAdapter.this.mHandler.removeCallbacks(KanojoInfoImgAdapter.this.mNotifyThread);
                KanojoInfoImgAdapter.this.mHandler.postDelayed(KanojoInfoImgAdapter.this.mNotifyThread, 300);
            }
        }
    }
}

File: src/main/java/com/goujer/barcodekanojo/adapter/KanojoInfoImgAdapter.kt
package com.goujer.barcodekanojo.adapter

import android.content.Context
import android.os.Handler
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.Gallery
import android.widget.ImageView
import com.goujer.barcodekanojo.core.util.DynamicImageCache
import jp.co.cybird.barcodekanojoForGAM.R
import jp.co.cybird.barcodekanojoForGAM.adapter.base.ObservableAdapter
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.util.*

class KanojoInfoImgAdapter(private val context: Context, private val mDic: DynamicImageCache) : BaseAdapter(), ObservableAdapter {
	private val galleryIconSize = context.resources.getDimension(R.dimen.kanojo_info_gallery_height).toInt()
	private var imgUrls: MutableList<String> = ArrayList<String>()
	private val mHandler = Handler()
	private val mNotifyThread = Runnable { notifyDataSetChanged() }
	private val mResourcesObserver = RemoteResourceManagerObserver(this, null as RemoteResourceManagerObserver?)
	private val mScope = MainScope()

	//fun setImgUrls(l: MutableList<String>) {
	//	imgUrls = l
	//	for (url in l) {
	//		mDic.requestBitmap(url)
	//	}
	//	notifyDataSetChanged()
	//	mHandler.removeCallbacks(mNotifyThread)
	//	mHandler.postDelayed(mNotifyThread, 200)
	//}

	fun addImgUrl(l: String) {
		imgUrls.add(l)
		notifyDataSetChanged()

		mDic.requestBitmap(l)
		mHandler.removeCallbacks(mNotifyThread)
		mHandler.postDelayed(mNotifyThread, 200)
	}

	fun destroy() {
		imgUrls.clear()
		mScope.cancel()
	}

	override fun getCount(): Int {
		return imgUrls.size
	}

	override fun getItem(position: Int): Any {
		return imgUrls[position]
	}

	override fun hasStableIds(): Boolean {
		return true
	}

	fun getItemUrl(position: Int): String {
		return imgUrls[position]
	}

	override fun getItemId(position: Int): Long {
		return position.toLong()
	}

	override fun getView(position: Int, view: View?, parent: ViewGroup): View? {
		val imageView: ImageView
		if (view == null) {
			imageView = ImageView(context)
			imageView.scaleType = ImageView.ScaleType.FIT_CENTER
			imageView.layoutParams = Gallery.LayoutParams(galleryIconSize, galleryIconSize)
		} else {
			imageView = view as ImageView
		}
		mScope.launch { mDic.loadBitmap(imageView, imgUrls[position], R.drawable.common_noimage_product, null) }
		return imageView
	}

	override fun removeObserver() {
		mHandler.removeCallbacks(mNotifyThread)
		mDic.deleteObserver(mResourcesObserver)
	}

	private inner class RemoteResourceManagerObserver private constructor() : Observer {
		/* synthetic */
		internal constructor(kanojoInfoImgAdapter: KanojoInfoImgAdapter?, remoteResourceManagerObserver: RemoteResourceManagerObserver?) : this() {}

		override fun update(observable: Observable, data: Any) {
			if (data != null) {
				mHandler.removeCallbacks(mNotifyThread)
				mHandler.postDelayed(mNotifyThread, 300)
			}
		}
	}

	companion object {
		private const val TAG = "KanojoInfoImgAdapter"
	}

	init {
		context.obtainStyledAttributes(R.styleable.KanojoInfoGallery).recycle()
		mDic.addObserver(mResourcesObserver)
	}
}